src/Routing/functions.php

```diff
@@ -0,0 +1,36 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+
+use Cake\Routing\Router;
+
+if (!function_exists('urlArray')) {
+    /**
+     * Returns an array URL from a route path string.
+     *
+     * @param string $path Route path.
+     * @param array $params An array specifying any additional parameters.
+     *   Can be also any special parameters supported by `Router::url()`.
+     * @return array URL
+     * @see \Cake\Routing\Router::pathUrl()
+     */
+    function urlArray(string $path, array $params = []): array
```

Comment:
Why not a name like `url()`, `route()` or `urlRoute()`? This method doesn't return an array, and can be used without any array parameters.

Reply:
My idea was to return an array here to be used further inside the application.
The array is the ideal data transport to be easily merged and modified, whereas the string seems to be the ideal input.
And they shouldn't be localized or finalized until the very end of the request, in order for the introspections and modifications to still be possible.

Reply:
This is purely opt-in right now, for those that want to use it instead of normal/classic strings.
I don't think we need to move to use those on core level or if you don't have the IDE support to actually be faster this way.

Reply:
Why not have another function too which returns string? Users can use whatever they prefer/need.

Reply:
So just url()? How would that function work?

Reply:
If `$url = urlArray('Users::index');` returns string, and app in subdir, it would return `'/subdir/users'`.

But if it then passed to `Router::url()` or `$this->redirect($url)`, subdir is applied again: `'/subdir/subdir/users'`.

So I would go with array.

Reply:
Yes, the final string is the worst format to use already, and continues to be so.
Unflexible and error prone when using it down the line, also not easily comparable where needed (as it is already final route string).

Reply:
@garas Yes, that can be issue, though if we allow additional params one can control whether then want the URL string with or without base.

Reply:
But rather than dealing with users reports of base getting prefixed twice, I can add my own `url()` function which wraps `Router::pathUrl()` :slightly_smiling_face:

Reply:
Feel free to add commits on top :)

Reply:
I am good with current state of PR. I'll put some more thought into the possibility of adding a function which returns string later.

Reply:
`Router::pathUrl()` and `urlArray()` now has different behavior.

`Router::pathUrl()` will reset prefix/plugin to `false`, but `urlArray()` will inherit prefix/plugin from current route.

`urlArray()` should better be just shortcut of `Router::pathUrl()`.

Reply:
I thought it behaved the same. Indeed, it should do so.

Reply:
I updated the code and tests.